# DATA SCIENCE FOR GOOD
## Data Preprocessing

Importing the necessary libraries.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
# import geopandas as gpd
import numpy as np
import os
from sklearn.preprocessing import LabelEncoder

Importing and reading the relevant datasets with pandas library. The links are given in a way that corresponds to the project structure.

In [3]:
interview_ds = pd.read_csv('../Dept_24-00013/24-00013_UOF_2008-2017_prepped.csv')
employment_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_employment/ACS_15_5YR_S2301_with_ann.csv')
edu_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_education-attainment/ACS_15_5YR_S1501_with_ann.csv')
edu25_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_education-attainment-over-25/ACS_15_5YR_B15003_with_ann.csv')
ownerhousing_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_owner-occupied-housing/ACS_15_5YR_S2502_with_ann.csv')
poverty_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_poverty/ACS_15_5YR_S1701_with_ann.csv')
income_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_income/ACS_15_5YR_S1903_with_ann.csv')
race_ds = pd.read_csv('../Dept_24-00013/24-00013_ACS_data/24-00013_ACS_race-sex-age/ACS_15_5YR_DP05_with_ann.csv')


In [4]:
interview_ds.head(10)

LOCATION_LONGITUDE LOCATION_LATITUDE INCIDENT_UNIQUE_IDENTIFIER  \
0                  X                 Y         PoliceUseOfForceID   
1       -93.27169927       45.00803595                          1   
2       -93.27169927       45.00803595                          2   
3       -93.27169927       45.00803595                          3   
4       -93.24701977       44.95974418                          4   
5       -93.24701977       44.95974418                          5   
6       -93.24701977       44.95974418                          6   
7       -93.27314066       44.98080828                          7   
8       -93.27314066       44.98080828                          8   
9       -93.27310907       44.97918668                          9   

  INCIDENT_DATE  INCIDENT_TIME SUBJECT_INJURY TYPE_OF_FORCE_USED  \
0  ResponseDate            NaN  SubjectInjury    ForceTypeDetail   
1    2008-01-01  01:15:31.000Z             -1    BodilyForceType   
2    2008-01-01  01:15:31.000Z             -1    BodilyForceType   
3    2008-01-01  01:15:31.000Z             -1    BodilyForceType   
4    2008-01-01  01:27:27.000Z              0    BodilyForceType   
5    2008-01-01  01:27:27.000Z              0    BodilyForceType   
6    2008-01-01  01:27:27.000Z              0      TaserDeployed   
7    2008-01-01  01:29:47.000Z            NaN    BodilyForceType   
8    2008-01-01  01:29:47.000Z            NaN    BodilyForceType   
9    2008-01-01  01:40:02.000Z            NaN    BodilyForceType   

  WEAPON_OR_TOOL_USED SUBJECT_RACE SUBJECT_GENDER SUBJECT_AGE  \
0     ForceTypeAction         Race            Sex    EventAge   
1           Push Away        Black         Female          25   
2  Body Weight to Pin        Black         Female          25   
3             Punches        Black         Female          25   
4          Joint Lock        White           Male          27   
5  Body Weight to Pin        White           Male          27   
6        Firing Darts        White           Male          27   
7           Push Away        White           Male          39   
8  Body Weight to Pin        White           Male          39   
9          Joint Lock        White           Male          42   

      REASON_FOR_FORCE LOCATION_DISTRICT  
0     TypeOfResistance          Precinct  
1    Assaulted Officer                 2  
2    Assaulted Officer                 2  
3    Assaulted Officer                 2  
4  Commission of Crime                 3  
5  Commission of Crime                 3  
6  Commission of Crime                 3  
7  Commission of Crime                 1  
8  Commission of Crime                 1  
9               Tensed                 1


Creating a single dataset (new_df) from the existing ones (employment_ds, edu_ds, edu25_ds, ownerhousing_ds, poverty_ds, income_ds, race_ds).


In [5]:
new_df = pd.concat([employment_ds, edu_ds, edu25_ds, ownerhousing_ds, poverty_ds, income_ds, race_ds], axis=1, sort=False)
new_df.head()

GEO.id      GEO.id2  \
0                    Id          Id2   
1  1400000US27053000101  27053000101   
2  1400000US27053000102  27053000102   
3  1400000US27053000300  27053000300   
4  1400000US27053000601  27053000601   

                               GEO.display-label  \
0                                      Geography   
1  Census Tract 1.01, Hennepin County, Minnesota   
2  Census Tract 1.02, Hennepin County, Minnesota   
3     Census Tract 3, Hennepin County, Minnesota   
4  Census Tract 6.01, Hennepin County, Minnesota   

                                   HC01_EST_VC01  \
0  Total; Estimate; Population 16 years and over   
1                                           2481   
2                                           3856   
3                                           2676   
4                                           4114   

                                       HC01_MOE_VC01  \
0  Total; Margin of Error; Population 16 years an...   
1                                                253   
2                                                262   
3                                                197   
4                                                346   

                                       HC02_EST_VC01  \
0  Labor Force Participation Rate; Estimate; Popu...   
1                                               72.1   
2                                               65.4   
3                                               76.7   
4                                               77.9   

                                       HC02_MOE_VC01  \
0  Labor Force Participation Rate; Margin of Erro...   
1                                                4.6   
2                                                6.9   
3                                                5.0   
4                                                4.7   

                                       HC03_EST_VC01  \
0  Employment/Population Ratio; Estimate; Populat...   
1                                               66.2   
2                                               60.5   
3                                               70.2   
4                                               74.2   

                                       HC03_MOE_VC01  \
0  Employment/Population Ratio; Margin of Error; ...   
1                                                6.0   
2                                                6.8   
3                                                5.0   
4                                                4.9   

                                       HC04_EST_VC01  \
0  Unemployment rate; Estimate; Population 16 yea...   
1                                                8.2   
2                                                7.5   
3                                                8.5   
4                                                4.8   

                         ...                          \
0                        ...                           
1                        ...                           
2                        ...                           
3                        ...                           
4                        ...                           

                                          HC03_VC108  \
0  Percent; CITIZEN, VOTING AGE POPULATION - Citi...   
1                                               2082   
2                                               3435   
3                                               2538   
4                                               3741   

                                          HC04_VC108  \
0  Percent Margin of Error; CITIZEN, VOTING AGE P...   
1                                                (X)   
2                                                (X)   
3                                                (X)   
4                                                (X)   

                                          HC01_VC109  \
0  Estimate; CITIZEN, VOTING AGE POPULATION - Cit...   
1    

In [6]:
new_df.shape # Looking at the shape of the dataset.

(300, 2103)

In [7]:
new_df.columns # Looking into the columns as the information given.

Index(['GEO.id', 'GEO.id2', 'GEO.display-label', 'HC01_EST_VC01',
       'HC01_MOE_VC01', 'HC02_EST_VC01', 'HC02_MOE_VC01', 'HC03_EST_VC01',
       'HC03_MOE_VC01', 'HC04_EST_VC01',
       ...
       'HC03_VC108', 'HC04_VC108', 'HC01_VC109', 'HC02_VC109', 'HC03_VC109',
       'HC04_VC109', 'HC01_VC110', 'HC02_VC110', 'HC03_VC110', 'HC04_VC110'],
      dtype='object', length=2103)

Substituting the missing data with the NaN format.

In [8]:
new_df = new_df.where(new_df!='(X)', None)
new_df.head(2)

GEO.id      GEO.id2  \
0                    Id          Id2   
1  1400000US27053000101  27053000101   

                               GEO.display-label  \
0                                      Geography   
1  Census Tract 1.01, Hennepin County, Minnesota   

                                   HC01_EST_VC01  \
0  Total; Estimate; Population 16 years and over   
1                                           2481   

                                       HC01_MOE_VC01  \
0  Total; Margin of Error; Population 16 years an...   
1                                                253   

                                       HC02_EST_VC01  \
0  Labor Force Participation Rate; Estimate; Popu...   
1                                               72.1   

                                       HC02_MOE_VC01  \
0  Labor Force Participation Rate; Margin of Erro...   
1                                                4.6   

                                       HC03_EST_VC01  \
0  Employment/Population Ratio; Estimate; Populat...   
1                                               66.2   

                                       HC03_MOE_VC01  \
0  Employment/Population Ratio; Margin of Error; ...   
1                                                6.0   

                                       HC04_EST_VC01  \
0  Unemployment rate; Estimate; Population 16 yea...   
1                                                8.2   

                         ...                          \
0                        ...                           
1                        ...                           

                                          HC03_VC108  \
0  Percent; CITIZEN, VOTING AGE POPULATION - Citi...   
1                                               2082   

                                          HC04_VC108  \
0  Percent Margin of Error; CITIZEN, VOTING AGE P...   
1                                               None   

                                          HC01_VC109  \
0  Estimate; CITIZEN, VOTING AGE POPULATION - Cit...   
1                                               1058   

                                          HC02_VC109  \
0  Margin of Error; CITIZEN, VOTING AGE POPULATIO...   
1                                                126   

                                          HC03_VC109  \
0  Percent; CITIZEN, VOTING AGE POPULATION - Citi...   
1                                               50.8   

                                          HC04_VC109  \
0  Percent Margin of Error; CITIZEN, VOTING AGE P...   
1                                                4.9   

                                          HC01_VC110  \
0  Estimate; CITIZEN, VOTING AGE POPULATION - Cit...   
1                                               1024   

                                          HC02_VC110  \
0  Margin of Error; CITIZEN, VOTING AGE POPULATIO...   
1                                                152   

                                          HC03_VC110  \
0  Percent; CITIZEN, VOTING AGE POPULATION - Citi...   
1                                               49.2   

                                          HC04_VC110  
0  Percent Margin of Error; CITIZEN, VOTING AGE P...  
1                                                4.9  

[2 rows x 2103 columns]

In [9]:
new_df.columns = new_df.iloc[0] # Replacing the existing headers with the relevant columns' names (those of the first row).
new_df.drop([0], axis=0, inplace=True) # dropping the first line of the dataset as it contains the headers
new_df.head()

0                    Id          Id2  \
1  1400000US27053000101  27053000101   
2  1400000US27053000102  27053000102   
3  1400000US27053000300  27053000300   
4  1400000US27053000601  27053000601   
5  1400000US27053000603  27053000603   

0                                      Geography  \
1  Census Tract 1.01, Hennepin County, Minnesota   
2  Census Tract 1.02, Hennepin County, Minnesota   
3     Census Tract 3, Hennepin County, Minnesota   
4  Census Tract 6.01, Hennepin County, Minnesota   
5  Census Tract 6.03, Hennepin County, Minnesota   

0 Total; Estimate; Population 16 years and over  \
1                                          2481   
2                                          3856   
3                                          2676   
4                                          4114   
5                                          2227   

0 Total; Margin of Error; Population 16 years and over  \
1                                                253     
2                                                262     
3                                                197     
4                                                346     
5                                                119     

0 Labor Force Participation Rate; Estimate; Population 16 years and over  \
1                                               72.1                       
2                                               65.4                       
3                                               76.7                       
4                                               77.9                       
5                                               77.6                       

0 Labor Force Participation Rate; Margin of Error; Population 16 years and over  \
1                                                4.6                              
2                                                6.9                              
3                                                5.0                              
4                                                4.7                              
5                                                4.7                              

0 Employment/Population Ratio; Estimate; Population 16 years and over  \
1                                               66.2                    
2                                               60.5                    
3                                               70.2                    
4                                               74.2                    
5                                               76.0                    

0 Employment/Population Ratio; Margin of Error; Population 16 years and over  \
1                                                6.0                           
2                                                6.8                           
3                                                5.0                           
4                                                4.9                           
5                                                4.9                           

0 Unemployment rate; Estimate; Population 16 years and over  \
1                                                8.2          
2                                                7.5          
3                                                8.5          
4                                                4.8          
5                                                2.1          

0                                                ...                                                  \
1                                                ...                                                   
2                                                ...                                                   
3                                                ...                                                   
4                                                ...                                                   
5              

Checking the number of missing values in the dataset, considering that the maximum number is 300.

In [10]:
new_df.isnull().sum(axis=0)

0
Id                                                                                                                                                                                       0
Id2                                                                                                                                                                                      0
Geography                                                                                                                                                                                0
Total; Estimate; Population 16 years and over                                                                                                                                            0
Total; Margin of Error; Population 16 years and over                                                                                                                                     0
Labor Force Participation Rate; Estimate; Population 16 years a

In [11]:
new_df.isnull().sum(axis=0).max()

299

Dropping all the columns that have more than 200 missing values.

In [12]:
new_df1 = new_df.dropna(axis=1, thresh=200)
new_df1.head()

0                    Id          Id2  \
1  1400000US27053000101  27053000101   
2  1400000US27053000102  27053000102   
3  1400000US27053000300  27053000300   
4  1400000US27053000601  27053000601   
5  1400000US27053000603  27053000603   

0                                      Geography  \
1  Census Tract 1.01, Hennepin County, Minnesota   
2  Census Tract 1.02, Hennepin County, Minnesota   
3     Census Tract 3, Hennepin County, Minnesota   
4  Census Tract 6.01, Hennepin County, Minnesota   
5  Census Tract 6.03, Hennepin County, Minnesota   

0 Total; Estimate; Population 16 years and over  \
1                                          2481   
2                                          3856   
3                                          2676   
4                                          4114   
5                                          2227   

0 Total; Margin of Error; Population 16 years and over  \
1                                                253     
2                                                262     
3                                                197     
4                                                346     
5                                                119     

0 Labor Force Participation Rate; Estimate; Population 16 years and over  \
1                                               72.1                       
2                                               65.4                       
3                                               76.7                       
4                                               77.9                       
5                                               77.6                       

0 Labor Force Participation Rate; Margin of Error; Population 16 years and over  \
1                                                4.6                              
2                                                6.9                              
3                                                5.0                              
4                                                4.7                              
5                                                4.7                              

0 Employment/Population Ratio; Estimate; Population 16 years and over  \
1                                               66.2                    
2                                               60.5                    
3                                               70.2                    
4                                               74.2                    
5                                               76.0                    

0 Employment/Population Ratio; Margin of Error; Population 16 years and over  \
1                                                6.0                           
2                                                6.8                           
3                                                5.0                           
4                                                4.9                           
5                                                4.9                           

0 Unemployment rate; Estimate; Population 16 years and over  \
1                                                8.2          
2                                                7.5          
3                                                8.5          
4                                                4.8          
5                                                2.1          

0                                                ...                                                  \
1                                                ...                                                   
2                                                ...                                                   
3                                                ...                                                   
4                                                ...                                                   
5              

Deleting the duplicating columns.

In [13]:
new_df1 = new_df1.T.drop_duplicates().T

Filling the missing data with the 'ffill' method.

In [14]:
new_df1 = new_df1.fillna(method='ffill')

Looking into the datatypes of the dataset. 

In [15]:
 new_df1.dtypes

0
Id                                                                                                                                                                                     object
Id2                                                                                                                                                                                    object
Geography                                                                                                                                                                              object
Total; Estimate; Population 16 years and over                                                                                                                                          object
Total; Margin of Error; Population 16 years and over                                                                                                                                   object
Labor Force Participation Rate; Estimate; Popula

## Looking into the second dataset.
 

In [16]:
interview_ds.shape # Checking the shape of the dataset

(25802, 13)

In [17]:
interview_ds.columns # Checking the columns for the information available

Index(['LOCATION_LONGITUDE', 'LOCATION_LATITUDE', 'INCIDENT_UNIQUE_IDENTIFIER',
       'INCIDENT_DATE', 'INCIDENT_TIME', 'SUBJECT_INJURY',
       'TYPE_OF_FORCE_USED', 'WEAPON_OR_TOOL_USED', 'SUBJECT_RACE',
       'SUBJECT_GENDER', 'SUBJECT_AGE', 'REASON_FOR_FORCE',
       'LOCATION_DISTRICT'],
      dtype='object')

Checking the number of missing values per column considering that the total number of values is 25802. 

In [18]:
interview_ds.isnull().sum(axis=0) # Checking the null values per column

LOCATION_LONGITUDE               0
LOCATION_LATITUDE                0
INCIDENT_UNIQUE_IDENTIFIER       0
INCIDENT_DATE                    0
INCIDENT_TIME                    1
SUBJECT_INJURY                7659
TYPE_OF_FORCE_USED             957
WEAPON_OR_TOOL_USED              0
SUBJECT_RACE                   957
SUBJECT_GENDER                 957
SUBJECT_AGE                    958
REASON_FOR_FORCE               961
LOCATION_DISTRICT              957
dtype: int64

Dropping the columns containing more than 7000 of missing values (SUBJECT_INJURY).

In [19]:
inter = interview_ds.drop(['SUBJECT_INJURY'], axis=1)


Replacing the existing headers with the relevant ones (located in the first row of the dataset).

In [20]:
inter.columns = inter.iloc[0] # Replacing the existing headers with the relevant columns' names (those of the first row).
inter.drop([0], axis=0, inplace=True) # dropping the first line of the dataset as it contains the headers
inter.head()

0             X            Y PoliceUseOfForceID ResponseDate            NaN  \
1  -93.27169927  45.00803595                  1   2008-01-01  01:15:31.000Z   
2  -93.27169927  45.00803595                  2   2008-01-01  01:15:31.000Z   
3  -93.27169927  45.00803595                  3   2008-01-01  01:15:31.000Z   
4  -93.24701977  44.95974418                  4   2008-01-01  01:27:27.000Z   
5  -93.24701977  44.95974418                  5   2008-01-01  01:27:27.000Z   

0  ForceTypeDetail     ForceTypeAction   Race     Sex EventAge  \
1  BodilyForceType           Push Away  Black  Female       25   
2  BodilyForceType  Body Weight to Pin  Black  Female       25   
3  BodilyForceType             Punches  Black  Female       25   
4  BodilyForceType          Joint Lock  White    Male       27   
5  BodilyForceType  Body Weight to Pin  White    Male       27   

0     TypeOfResistance Precinct  
1    Assaulted Officer        2  
2    Assaulted Officer        2  
3    Assaulted Officer        2  
4  Commission of Crime        3  
5  Commission of Crime        3

Filling the missing value using the 'ffill' method.

In [21]:
inter = inter.fillna(method='ffill')

Transforming the categorical data into the numerical one using LabelEncoder.

In [22]:
l_e = LabelEncoder()
inter.ForceTypeDetail = l_e.fit_transform(inter.ForceTypeDetail)
inter.Race = l_e.fit_transform(inter.Race)
inter.Sex = l_e.fit_transform(inter.Sex)
inter.ForceTypeAction = l_e.fit_transform(inter.ForceTypeAction)
inter.TypeOfResistance = l_e.fit_transform(inter.TypeOfResistance)

In [23]:
inter.rename(columns={'Sex' : 'Male'}, inplace=True) # Changing the name of the column so that 1 = Male looks more obvious
data=inter
inter.head()

0             X            Y PoliceUseOfForceID ResponseDate            NaN  \
1  -93.27169927  45.00803595                  1   2008-01-01  01:15:31.000Z   
2  -93.27169927  45.00803595                  2   2008-01-01  01:15:31.000Z   
3  -93.27169927  45.00803595                  3   2008-01-01  01:15:31.000Z   
4  -93.24701977  44.95974418                  4   2008-01-01  01:27:27.000Z   
5  -93.24701977  44.95974418                  5   2008-01-01  01:27:27.000Z   

0  ForceTypeDetail  ForceTypeAction  Race  Male EventAge  TypeOfResistance  \
1                1               29     1     0       25                 0   
2                1                4     1     0       25                 0   
3                1               28     1     0       25                 0   
4                1               17     6     1       27                 4   
5                1                4     6     1       27                 4   

0 Precinct  
1        2  
2        2  
3        2  
4        3  
5        3

Transforming the data into binary format by using dummies.

In [24]:
d = [5,6,7,10,11]
for i in d:
    n = pd.get_dummies(inter[inter.columns[i]])
    inter = pd.concat([inter, n], axis=1)
inter.head()
    

X            Y PoliceUseOfForceID ResponseDate            NaN  \
1  -93.27169927  45.00803595                  1   2008-01-01  01:15:31.000Z   
2  -93.27169927  45.00803595                  2   2008-01-01  01:15:31.000Z   
3  -93.27169927  45.00803595                  3   2008-01-01  01:15:31.000Z   
4  -93.24701977  44.95974418                  4   2008-01-01  01:27:27.000Z   
5  -93.24701977  44.95974418                  5   2008-01-01  01:27:27.000Z   

   ForceTypeDetail  ForceTypeAction  Race  Male EventAge ...  10 11  12  13  \
1                1               29     1     0       25 ...   0  0   0   0   
2                1                4     1     0       25 ...   0  0   0   0   
3                1               28     1     0       25 ...   0  0   0   0   
4                1               17     6     1       27 ...   0  0   0   0   
5                1                4     6     1       27 ...   0  0   0   0   

   0  1  2  3  4  5  
1  0  0  1  0  0  0  
2  0  0  1  0  0  0  
3  0  0  1  0  0  0  
4  0  0  0  1  0  0  
5  0  0  0  1  0  0  

[5 rows x 86 columns]

Checking the type of the data in the dataset.

In [25]:
inter.dtypes

X                     object
Y                     object
PoliceUseOfForceID    object
ResponseDate          object
NaN                   object
ForceTypeDetail        int64
ForceTypeAction        int64
Race                   int64
Male                   int64
EventAge              object
TypeOfResistance       int64
Precinct              object
0                      uint8
1                      uint8
2                      uint8
3                      uint8
4                      uint8
5                      uint8
6                      uint8
7                      uint8
8                      uint8
0                      uint8
1                      uint8
2                      uint8
3                      uint8
4                      uint8
5                      uint8
6                      uint8
7                      uint8
8                      uint8
                       ...  
35                     uint8
36                     uint8
0                      uint8
1             

Changing the data type into the datetime format. 

In [26]:
inter.ResponseDate = pd.to_datetime(inter.ResponseDate, format='%Y-%m-%d')

In [31]:
k = inter.iloc[:,4]
k.rename('Time', inplace=True)
# inter = pd.concat(inter.drop(inter.columns[4], axis=1),k)
inter.head()

X            Y PoliceUseOfForceID ResponseDate            NaN  \
1  -93.27169927  45.00803595                  1   2008-01-01  01:15:31.000Z   
2  -93.27169927  45.00803595                  2   2008-01-01  01:15:31.000Z   
3  -93.27169927  45.00803595                  3   2008-01-01  01:15:31.000Z   
4  -93.24701977  44.95974418                  4   2008-01-01  01:27:27.000Z   
5  -93.24701977  44.95974418                  5   2008-01-01  01:27:27.000Z   

   ForceTypeDetail  ForceTypeAction  Race  Male EventAge ...  10 11  12  13  \
1                1               29     1     0       25 ...   0  0   0   0   
2                1                4     1     0       25 ...   0  0   0   0   
3                1               28     1     0       25 ...   0  0   0   0   
4                1               17     6     1       27 ...   0  0   0   0   
5                1                4     6     1       27 ...   0  0   0   0   

   0  1  2  3  4  5  
1  0  0  1  0  0  0  
2  0  0  1  0  0  0  
3  0  0  1  0  0  0  
4  0  0  0  1  0  0  
5  0  0  0  1  0  0  

[5 rows x 86 columns]

In [32]:
ind_col_names = [x for x in inter if x!= 'Race'] #setting the parameters, choosing the target for interview ds
ind_col_names1 = [x for x in new_df1 if x!= ''] #setting the parameters, choosing the target for the major table
X = 
y = 

from sklearn.model_selection import train_test_split
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.1, random_state=17)

5

## Data Analysis 

In [90]:
data.head()

0             X            Y PoliceUseOfForceID ResponseDate            NaN  \
1  -93.27169927  45.00803595                  1   2008-01-01  01:15:31.000Z   
2  -93.27169927  45.00803595                  2   2008-01-01  01:15:31.000Z   
3  -93.27169927  45.00803595                  3   2008-01-01  01:15:31.000Z   
4  -93.24701977  44.95974418                  4   2008-01-01  01:27:27.000Z   
5  -93.24701977  44.95974418                  5   2008-01-01  01:27:27.000Z   

0  ForceTypeDetail  ForceTypeAction  Race  Male EventAge  TypeOfResistance  \
1                1               29     1     0       25                 0   
2                1                4     1     0       25                 0   
3                1               28     1     0       25                 0   
4                1               17     6     1       27                 4   
5                1                4     6     1       27                 4   

0 Precinct  
1        2  
2        2  
3        2  
4        3  
5        3

In [93]:
data.ResponseDate[3]

str